### DISCUSSION TITLE
Fourth order polar approximation

### DISCUSSION THREAD

**Discussion by hjr** [XCSoar/XCSoar#806] [Ideas]
Certainly not a new idea. What are the pros and cons to this. CPU horsepower would certainly not be the barrier any more. Just a simple pic to illustrate the obvious advantage.
![PolarFourthOrder](https://user-images.githubusercontent.com/415029/147052829-0e94ec17-a576-46be-8363-6f3b31c8aa1e.jpg)
The lower speed and as well the typical kink at about beyond best l/d speeds are captured. A s2f optimizer should not recommend nonsense below min speed any more, as well as be more considerate on high speeds.

I did a rough google search without any good findings to literature on this topic.

Thx

**Comment by MaxKellermann** [comment]
CPU-wise, the terrain renderer is ten thousand times more expensive (or so). You are right, CPU is not a problem at all - and never was. Even those old iPaqs could have done that easily.

**Comment by iltis42** [comment]
Yes i agree, crunching two more terms for S2F should no longer be a problem. Almost any embedded system comes nowadays with a hardware floating point unit doing addition, multiplication or division incredibly fast, such as 150 nanoseconds.
Similar to the second order, also the math for the 4th order is not issue, and can be extended in a similar way, with a given 4th order approximation and sqrt(o) for the root of ballast overload, we had to crunch for ballasted Sink something like this:

![4thOrder](https://user-images.githubusercontent.com/34288016/151623399-2d157641-a6c2-4484-be29-a0aafabdc7b9.png)

**Comment by dssherrill** [comment]
Computing 4th order sink is easy.  But the inverse problem (computing 4th order MacCready STF) is hard.

With the quadratic model, you can immediately solve for the speed that corresponds to a given value of sink by using the quadratic formula everyone learns in grade school.    XCSoar does that now.

But for a higher order model, you must hunt for a solution iteratively.  That is also a trivial CPU load, unless you do it frequently.  So, how often does auto-MacCready update?  And how many other calculations also need this result?  Task optimizer, maybe?

And how fast is the the CPU?  Many people run XCSoar on very old hardware.

If the CPU burden is too high, an intermediate solution might be to have multiple quadratic models, one for low speed, one intermediate speeds, and one for high speed.  The speed ranges can overlap.  The solver would pick the model that is most nearly centered on the sink rate value, and then solve for the speed using the quadratic equation code in XCSoar today.

Unfortunately, with multiple models, when a small change in the sink value causes the solver to switch models, there is likely to be a discontinuity in the computed speed.

So, this would not be my first choice.  It would be better to find ways to reduced the CPU load of the iterative solution.

**Comment by Simsys** [comment]
Polynomials of higher order tend to oscillate. You quickly get a polar that has waves. At speeds outside the approximation, completely crazy values can appear. This has ugly consequences: The speed to fly then jumps ‘over the valleys’, which causes severe irritation for the pilot. You would have to define the polars very carefully or you need a tool that does this. This is anything but trivial.

A friend has implemented a different approach. He has attached four 2nd order polynomials to each other. The slope and descent match at the connection points. This can be implemented very easily and efficiently in terms of programming, including speed-to-fly determination. But even with this solution, jumps and unsteady behaviour of the speed to fly can easily occur.

Here are the polar diagrams (sinkrate and speet to fly) of the DG800b and ASG32
![image](https://github.com/XCSoar/XCSoar/assets/3678273/f9a15be3-3e12-4e8d-81f4-8a1e60eac511)
![image](https://github.com/XCSoar/XCSoar/assets/3678273/c531ff3a-d79b-4a86-83ca-f382b0d7c68f)
![image](https://github.com/XCSoar/XCSoar/assets/3678273/b9541ee2-38a1-4709-b287-687058525853)
![image](https://github.com/XCSoar/XCSoar/assets/3678273/c6e5c797-3888-4e2e-9308-92f38b1f1f7d)

Such a solution is quite difficult to determine the correct approximate values.

On the other hand, I ask myself the question: what is the point? How exactly do we know and need the polars? When I look at the fidgeting of the speed to fly indicator in flight.

Uncertainties:
- Centre of gravity position
- Insect covering
- Load including water ballast
- Condition of the aircraft
- Behaviour of the glider in gusts

I doubt the benefit of such optimisations.

**Reply by GuillermoHazebrouck** [reply]
I have been working with fourth order polynomial in the CD(CL) domain, and what I have observed is pretty good actually. The gliding slope is much more accurate.
The issue with the speed to fly is that the gliding slope becomes insensitive to it as the MC value decreases (sink).

**Comment by lordfolken** [comment]
In the end we want a STF that is useful for cruise and final glide. 
I think we have to look at the mc value's rate of change and see whether those are realistic. 
You can only so and so fast correct a speed that is displayed to you. Say 10km per second. 
Potentially also use a longer period of time to assess the current air-mass, for example 10 seconds.

**Comment by Plantain** [comment]
Why bother with polars and poor approximations like this at all? Why
not just a lookup table of the speed/sink every 5 kph?

**Comment by GuillermoHazebrouck** [comment]
I've been studying this subject, and what I see (for 6 common glider models) is that working with four order polynomial (on the CL-CD domain) results in a superior fitting for the whole speed range. This is good for the range estimate mainly.
For the speed to fly, what I have observed is that the slope is very insensitive to the airspeed around the optimal point. With most models it seems you can fly +-10km/h or more around the optimal spend, and see almost no difference in the slope. This effect even accentuates with increased sink. The slope can be affected by the sink in a much higher degree. See example plots here for the Duo.

![Duo-Discus_Slope](https://github.com/user-attachments/assets/0d8f4683-fcea-4374-acde-6ce54de45f13)
![Duo-Discus_Polar](https://github.com/user-attachments/assets/91561e22-1fbf-424a-af73-c957f6557fc3)

**Comment by bomilkar** [comment]
I'd like to add the **physical aspect**:
The quadratic equation (which we have now) is more than just a simple fit. It reflects physics: drag and lift.
So what physical aspect do you have in mind when you add the cubic term?
If there is no physical effect motivating a cubic term then I would rather vote for a lookup table as suggested by @Plantain (perhaps not equidistant) .

**Reply by bomilkar** [reply]
> I'd like to add the **physical aspect**: The quadratic equation (which we have now) is more than just a simple fit. It reflects physics: drag and lift. So what physical aspect do you have in mind when you add the cubic term? If there is no physical effect motivating a cubic term then I would rather vote for a lookup table as suggested by @Plantain (perhaps not equidistant) .

Perhaps I should take back my vote in favor of lookup table.
Taking wing load, bugs, flap position into account is certainly possible with lookup tables: just add a dimension for every additional parameter. But is that practical??? Certainly not!

On a similar token:
calculating the polare form a quadratic polynom for different parameters (wing load, bugs, flap position) is obvious just because a quadratic polynom reflects the physics. How would these parameters apply to a cubic polynom???

**Reply by GuillermoHazebrouck** [reply]
The concept of a quadratic polar comes originally from the potential flow theory after some assumptions and simplifications and it is only valid for the induced drag component (i.e. pressure drag, and only on the wing). It is a nice theoretical result that comes close to most practical cases. But for the skin drag, which amounts a large portion of the total drag at low CL (i.e. higher speed), there is no rule at all that says it should behave quadratically and the physics behind is complex, I dear to say almost beyond pure theoretical predictions. For modern gliders, a simple quadratic curve is not a good approach because you cannot tune it to match the actual curve along a large range of CL.
Now, the rough (bug) polar and anything in between the clean and rough polars is not given anywhere, and that's I think the trickiest part of all this. I would imagine the rough polar as a more quadratic-like curve, because that's what can be see on some wind tunnel results (Abbott, Theory of wing sections). So a full quadratic polar could be okay for a rough surface. But in general, I have observed that cubic polars give a much better representation for the whole range and cases.
My plots are in the generic CD/CL domain, and in my opinion they should remain there in the program too to allow for recombination of states, but I think (I should recheck) that this is not the case in XCSoar.
I know I am nobody in the eye of the gliding community (and XCSoar) and you can disregard all I say, but if you follow my profile you'll see at least I am talking out of real experience on this topic.

**Comment by dssherrill** [comment]
I've been thinking about this a lot since my first post, working on it slowly, and talking to other pilots.  I now have a Python app that can compute speed-to-fly using the actual polar curve as input.

You can see from the first image below that the quadratic model (Degree = 2) underestimates STF quite a bit at the speeds we are most interested in.  Some doubt about these results arises because the shape of the jump around MC=1 changes so dramatically with the order of the model.  But I am convinced that the jump is real because you get the same result using the graphical method of finding MacCready speeds.  It is all the result of the straight (or low curvature) section of the polar curve.  For the ASW 28 curve shown, flat spot lies between 100 and 150 kph. 

So the differences are pretty big, but alas, I'm starting to think that it not very significant because (1) we fly using block speeds instead of slavishly obeying the STF from the flight director display, and (2) Bernard Eckey points out that flying at the wrong speed (say 25% error on MacCready setting) does not change your average speed very much.

As a third point, other pilots have pointed out that this only shows up on modern (laminar flow) airfoils on gliders without flaps.  So it is not an issue for the Discus (too old) but is an issue for the Discus 2.  (I can't vouch for that because I have not seen the polar curve for either glider.) 

Then again, if the polar for a flapped glider is a composite of the curves for each flap setting, then I see no reason for it to fit a quadratic curve, either.  Once again, I have not seen the polars.

The only relevant polars I have been able to digitize are the ASW 28 and the Duo Discus T.  The polars must be digitized with considerable care or the high-order fits fall apart.  And then you must hope that the glider manufacturer put the same care into the published polar chart you are about to digitize.

The software is a bit fragile and buggy, but you can play with it here:  
https://polarproject-production.up.railway.app/

You can review the code at the link below, but please be gentle in your scorn.  I am the first to admit that it is not pretty.  This is my first Python project, but that does not excuse the obvious lack of planning in the code.  Suggestions for improvement are welcome.
https://github.com/dssherrill/PolarProject

Cheers,
   ...david

<img width="1277" height="1086" alt="image" src="https://github.com/user-attachments/assets/6bd2e6b0-c89f-4c8e-89ff-a390e72340f7" />

<img width="600" height="400" alt="newplot" src="https://github.com/user-attachments/assets/fe8018f7-fcda-4a1e-8269-578fa066bf12" />

**Reply by GuillermoHazebrouck** [reply]
The first graph does not indicate the resulting slopes, which also matters. If you plot the slope curves (like I did above for the Duo), you will see they become very flat around the speed to fly, which means that it does not really matter if you fly faster or slower within a certain speed range. I will try to digitalize the curve and process it with my Scilab script to see what I get.

**Comment by GuillermoHazebrouck** [comment]
Yes I agree with you that this should not be a big concern. For most gliders the best slope is not too sensitive on the speed... you can fly more or less than 10km/h around the optimal point and still you'll be doing very good.
Also indeed, the polaris that the manufacturer provides are for general purpose and who knows what's the real agreement with reality. Even of they were obtained on wind tunnel, there might be discrepancies.
But I still really think that taking a 3rd order polynomial won't hurt, since errors can cumulate. From the computational view, it only requires two extra multiplications and a sum to accomplish. And if you want, you can still make it work as a quadratic curve.
For me it is still better to work on the CL/CD domain, to split the clean and rough components there and to recompose based on the scaling of the parasite drag only. Induced wing drag should never be scaled.

**Comment by yorickreum** [comment]
From my quick check with a DG 500 polar, neither a 3rd nor 4th order polynomial describes the datapoints very well ... in both cases there is clearly systematics in the residuals. At this point we should properly just interpolate / Taylor fit with a high-order polynomial. Or understand the underlying physics better to obtain a better formula...

**Reply by GuillermoHazebrouck** [reply]
The physics is already in the curve, but the Speed/Sink variables are masking some details. There is a whole misunderstanding in thinking that the polar should be good with a quadratic function. In fact, the only thing that is known to be almost quadratic by nature is the relationship between the induced drag coefficient and the lift coefficient. The remaining part of the drag is not necessarily quadratic and it also depends heavily on the Raynods number and the state of the surface (which is not very quantifiable). These two conditions also correlate very well with states of the boundary layer. A lot of aerodynamicists have tried to find good theoretical explanations for the boundary layer behavior to try to predict it, but there is not a single successful theory (and there will never be). So you will not find a formula other than one made by an empirical correlation.
What I think is a consistent approach is to revert the polar to CL/CD. By doing this, you also remove the dynamic pressure from the equation, which is quadratic to the air speed and adds unnecessary harmonics. Then you can also remove the induced quadratic component and interpolate only the remaining part, which is also the only part that can be affected by the bugs. So by doing that you tackle two problems at once. This is what I did in my program at least. The Scilab script I used for this is there in the project files...
